File: packages/client-android/src/test/kotlin/dev/poppy/android/snapshot/SnapshotRenderer.kt
/*
 * Render a typed PoppyDocument to the stable plain-text format committed
 * alongside each valid corpus case as `snapshot.android.txt`.
 *
 * Mirrors the iOS renderer at packages/client-ios/Tests/PoppyTests/
 * SnapshotRenderer.swift verbatim — same indentation, same key order, same
 * quoting rules. Cross-platform reviewers can diff snapshot.android.txt
 * against snapshot.ios.txt directly to spot wire-format-level divergences.
 *
 * Why a custom dumper instead of Compose's printToString()?
 *
 *   1. Compose's semantics tree is built for accessibility, not visual
 *      styling. It captures Text content, ContentDescription, Role, and
 *      TestTag — but NOT colors, sizes, weights, spacing, padding, or
 *      alignment. A snapshot that doesn't include the styling tokens can't
 *      catch the most common kind of regression (a token applied wrong).
 *   2. Plain layout containers like Column/Row don't emit semantic nodes
 *      unless they carry semantics modifiers — so structural hierarchy is
 *      invisible too.
 *   3. Test-time image stubs leak into the snapshot (a placeholder Text
 *      node with the URL appears where an Image should be).
 *
 * Walking the typed document tree avoids all three. The snapshot reflects
 * what the schema actually says, which is what reviewers care about.
 *
 * Format example (for case 015-kitchen-sink):
 *
 *   PoppyDocument version="0.1"
 *     Stack id="screen-root" axis=vertical padding=lg spacing=md alignment=stretch
 *       Stack axis=horizontal spacing=sm alignment=center
 *         Image url="..." alt="User avatar" width=48 height=48 fit=cover
 *         Stack axis=vertical spacing=xs
 *           Text size=lg weight=bold value="Greg"
 *           Text color=secondary size=sm value="Signed in"
 *       Text color=primary value="You have 3 unread notifications."
 *       Stack axis=horizontal spacing=sm alignment=end
 *         Button label="Dismiss" action.type=navigate action.uri="poppy://notifications/dismiss-all"
 *         Button label="View all" action.type=navigate action.uri="poppy://notifications"
 */

package dev.poppy.android.snapshot

import dev.poppy.android.Action
import dev.poppy.android.Alignment
import dev.poppy.android.Axis
import dev.poppy.android.Color
import dev.poppy.android.Component
import dev.poppy.android.Fit
import dev.poppy.android.PoppyDocument
import dev.poppy.android.Size
import dev.poppy.android.Spacing
import dev.poppy.android.Weight

internal object SnapshotRenderer {

    fun render(document: PoppyDocument): String {
        val lines = mutableListOf<String>()
        lines += "PoppyDocument version=\"${document.version}\""
        renderComponent(document.root, depth = 1, into = lines)
        return lines.joinToString("\n") + "\n"
    }

    private fun renderComponent(component: Component, depth: Int, into: MutableList<String>) {
        when (component) {
            is Component.Stack -> {
                into += line(depth, "Stack", stackAttrs(component))
                for (child in component.children) renderComponent(child, depth + 1, into)
            }
            is Component.Text -> into += line(depth, "Text", textAttrs(component))
            is Component.Image -> into += line(depth, "Image", imageAttrs(component))
            is Component.Button -> into += line(depth, "Button", buttonAttrs(component))
        }
    }

    // --- Per-component attribute lists ---------------------------------------
    //
    // Order is fixed for stability and parity with iOS's SnapshotRenderer.swift.
    // Optional fields are emitted only when set.

    private fun stackAttrs(s: Component.Stack): List<Pair<String, String>> = buildList {
        s.id?.let { add("id" to quoted(it)) }
        add("axis" to s.axis.wire())
        s.padding?.let { add("padding" to it.wire()) }
        s.spacing?.let { add("spacing" to it.wire()) }
        s.alignment?.let { add("alignment" to it.wire()) }
    }

    private fun textAttrs(t: Component.Text): List<Pair<String, String>> = buildList {
        t.id?.let { add("id" to quoted(it)) }
        t.color?.let { add("color" to it.wire()) }
        t.size?.let { add("size" to it.wire()) }
        t.weight?.let { add("weight" to it.wire()) }
        add("value" to quoted(t.value))
    }

    private fun imageAttrs(i: Component.Image): List<Pair<String, String>> = buildList {
        i.id?.let { add("id" to quoted(it)) }
        add("url" to quoted(i.url))
        add("alt" to quoted(i.alt))
        i.width?.let { add("width" to it.toString()) }
        i.height?.let { add("height" to it.toString()) }
        i.fit?.let { add("fit" to it.wire()) }
    }

    private fun buttonAttrs(b: Component.Button): List<Pair<String, String>> = buildList {
        b.id?.let { add("id" to quoted(it)) }
        add("label" to quoted(b.label))
        when (val a = b.action) {
            is Action.Navigate -> {
                add("action.type" to "navigate")
                add("action.uri" to quoted(a.uri))
            }
        }
    }

    // --- Formatting helpers --------------------------------------------------

    private fun line(depth: Int, head: String, attrs: List<Pair<String, String>>): String {
        val indent = "  ".repeat(depth)
        val body = attrs.joinToString(" ") { (k, v) -> "$k=$v" }
        return if (body.isEmpty()) "$indent$head" else "$indent$head $body"
    }

    private fun quoted(s: String): String {
        val escaped = s.replace("\\", "\\\\").replace("\"", "\\\"")
        return "\"$escaped\""
    }

    // --- Wire-format names for token enums -----------------------------------
    //
    // The schema uses lowercase strings; our enum constants are uppercase. The
    // mapping is consistently `name.lowercase()` for every v0.1 token, which
    // also matches what kotlinx.serialization emits via @SerialName. We use
    // `.lowercase()` rather than reflecting on @SerialName for simplicity —
    // a deviation here would be caught by the corpus snapshot diff against
    // iOS's output.

    private fun Axis.wire(): String = name.lowercase()
    private fun Spacing.wire(): String = name.lowercase()
    private fun Size.wire(): String = name.lowercase()
    private fun Color.wire(): String = name.lowercase()
    private fun Weight.wire(): String = name.lowercase()
    private fun Alignment.wire(): String = name.lowercase()
    private fun Fit.wire(): String = name.lowercase()
}

File: packages/client-ios/Tests/PoppyTests/SnapshotRenderer.swift
// Render a typed PoppyDocument to the stable plain-text format committed
// alongside each valid corpus case as `snapshot.ios.txt`.
//
// Why a custom dumper instead of ViewInspector's tree dump?
//
// 1. We control the format completely — line-stable, no implicit AnyView
//    noise, no stripping of internal SwiftUI machinery.
// 2. The output mirrors `snapshot.web.html`'s spirit: every component on
//    its own line(s) in document order, each line capturing the structural
//    shape (axis, fit, padding, spacing, alignment, etc.) reviewers compare.
// 3. The output is identical regardless of macOS / Xcode version, so
//    snapshots don't churn between CI matrix entries.
//
// Format example (for case 015-kitchen-sink):
//
//   PoppyDocument version="0.1"
//   Stack id="screen-root" axis=vertical padding=lg spacing=md alignment=stretch
//     Stack axis=horizontal spacing=sm alignment=center
//       Image url="…" alt="User avatar" width=48 height=48 fit=cover
//       Stack axis=vertical spacing=xs
//         Text size=lg weight=bold value="Greg"
//         Text size=sm color=secondary value="Signed in"
//     Text color=primary value="You have 3 unread notifications."
//     Stack axis=horizontal spacing=sm alignment=end
//       Button label="Dismiss" action.type=navigate action.uri="poppy://notifications/dismiss-all"
//       Button label="View all" action.type=navigate action.uri="poppy://notifications"
//
// The format is deliberately simple: indentation = depth, key=value pairs
// in deterministic order, double-quoted strings. No JSON / YAML — the snapshot
// is a human-readable diff target, not a machine-consumed format.

import Foundation
@testable import Poppy

enum SnapshotRenderer {

    static func render(_ document: PoppyDocument) -> String {
        var lines: [String] = []
        lines.append("PoppyDocument version=\"\(document.version)\"")
        renderComponent(document.root, depth: 1, into: &lines)
        return lines.joined(separator: "\n") + "\n"
    }

    private static func renderComponent(_ component: Component, depth: Int, into lines: inout [String]) {
        switch component {
        case .stack(let s):
            lines.append(line(depth: depth, head: "Stack", attrs: stackAttrs(s)))
            for child in s.children {
                renderComponent(child, depth: depth + 1, into: &lines)
            }
        case .text(let t):
            lines.append(line(depth: depth, head: "Text", attrs: textAttrs(t)))
        case .image(let i):
            lines.append(line(depth: depth, head: "Image", attrs: imageAttrs(i)))
        case .button(let b):
            lines.append(line(depth: depth, head: "Button", attrs: buttonAttrs(b)))
        }
    }

    // MARK: - Per-component attribute lists
    //
    // Order is fixed for stability. Optional fields are emitted only when set.

    private static func stackAttrs(_ s: Stack) -> [(String, String)] {
        var a: [(String, String)] = []
        if let id = s.id { a.append(("id", quoted(id))) }
        a.append(("axis", s.axis.rawValue))
        if let p = s.padding { a.append(("padding", p.rawValue)) }
        if let sp = s.spacing { a.append(("spacing", sp.rawValue)) }
        if let al = s.alignment { a.append(("alignment", al.rawValue)) }
        return a
    }

    private static func textAttrs(_ t: Text) -> [(String, String)] {
        var a: [(String, String)] = []
        if let id = t.id { a.append(("id", quoted(id))) }
        if let c = t.color { a.append(("color", c.rawValue)) }
        if let s = t.size { a.append(("size", s.rawValue)) }
        if let w = t.weight { a.append(("weight", w.rawValue)) }
        a.append(("value", quoted(t.value)))
        return a
    }

    private static func imageAttrs(_ i: Image) -> [(String, String)] {
        var a: [(String, String)] = []
        if let id = i.id { a.append(("id", quoted(id))) }
        a.append(("url", quoted(i.url)))
        a.append(("alt", quoted(i.alt)))
        if let w = i.width { a.append(("width", trimDouble(w))) }
        if let h = i.height { a.append(("height", trimDouble(h))) }
        if let f = i.fit { a.append(("fit", f.rawValue)) }
        return a
    }

    private static func buttonAttrs(_ b: Button) -> [(String, String)] {
        var a: [(String, String)] = []
        if let id = b.id { a.append(("id", quoted(id))) }
        a.append(("label", quoted(b.label)))
        switch b.action {
        case .navigate(let n):
            a.append(("action.type", "navigate"))
            a.append(("action.uri", quoted(n.uri)))
        }
        return a
    }

    // MARK: - Formatting helpers

    private static func line(depth: Int, head: String, attrs: [(String, String)]) -> String {
        let indent = String(repeating: "  ", count: depth)
        let body = attrs.map { "\($0.0)=\($0.1)" }.joined(separator: " ")
        return body.isEmpty ? "\(indent)\(head)" : "\(indent)\(head) \(body)"
    }

    private static func quoted(_ s: String) -> String {
        let escaped = s
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "\"", with: "\\\"")
        return "\"\(escaped)\""
    }

    private static func trimDouble(_ d: Double) -> String {
        // Render integral doubles (200.0) as "200", non-integral (1.5) as "1.5".
        if d.rounded() == d {
            return String(Int(d))
        }
        return String(d)
    }
}
